_posts/2019-05-10-why-projects-need-codenames.markdown

```diff
@@ -0,0 +1,47 @@
+
+**Language is important**, and not just for communication but for shaping our thoughts and assumptions. We frequently find ourselves debating which system should serve as the authority for a given domain model or where to implement a new feature, and project names that overlap with these topics (e.g., `search`, `images`, `suggestions`) would predispose us to certain decisions. Code names free us to focus on the architectural and organizational merits instead.
+
+With this in mind, how should you choose a naming scheme?
+
+### Rules for a project naming scheme
+
+* Avoid implicit value judgments like "new," "next," or "modern." We've all witnessed today's hot project become next year's unloved albatross.
+* Choose a code name scheme that isn't directly related to your technology or business. A flower business using flower names is cute, but breaks down when you want to build a feature that _actually_ is about tulips.
+* Everyone should be able to participate, so avoid industry lingo or anything obscure. Funny story: I work at an art start-up but don't know enough artists to name my projects that way. I struggle even more to pronounce the few projects that are.
+* There should be lots of choices. You'll regret choosing to name your projects after "decathlon sports" or "days of the week" when your team inevitably transitions to microservices.
+* Names should be unique within a company and--ideally--beyond the company's scope. You wouldn't want your `marketing-site` repository to conflict with a contributor's `marketing-site` repository from another organization. Github projects, S3 buckets, Heroku application names, and published libraries all benefit from being globally unique.
+* Bonus points for choosing a rich enough scheme that names can be subtly related to each system's function. At Artsy, we tend to name projects with physics terms, so I'd code-name a shipping service _Weight_.
+
+Example code name schemes: animals, movies, sea creatures, cartoon/TV characters, woodworking tools, celebrities (but get legal advice before borrowing living individuals' names). [Ubuntu](https://wiki.ubuntu.com/DevelopmentCodeNames) gets extra mileage by combining adjectives with animal names. [Apple](https://en.wikipedia.org/wiki/List_of_Apple_codenames) likes cats. [Google](https://en.wikipedia.org/wiki/Android_%28operating_system%29) likes dessert.
+
+
+### Finally
+
+These days, I've embraced Artsy's use of physics terms for project names. There continues to be some head-scratching as new engineers navigate codebases and encounter these terms or just struggle to identify the system responsible for a given site/feature. (This isn't shocking, since code names were probably adapted from more secretive environments like the military.) Once you have more than a few names in play, I suggest creating a simple glossary that maps projects' code names to descriptions, URL destinations, hosting environments, and teams.
```

Comment:
This is great – I like that you call out this downside. I think you could lean into it more, something like:

> [..] just struggle to identify the system responsible for a given site/feature. As old systems are retired and their codenames fall out of common parlance, a glossary becomes essential. Once you have more than a few names in play [...]

Or something like that, maybe.